WIP: example on morphing subject volume grid to fsaverage (GSoC)

TommyClausner opened the pull request:
example on morphing subject grid to fsaverage (GSoC)

jona-sassenhagen commented on examples/inverse/plot_morph_stc_to_fsaverage_dipy.py:
```diff
@@ -0,0 +1,231 @@
+plt.imshow(t1_s[slice_sel, :, :] + stc_sdr_affine[slice_sel, :, :])
+plt.axis('off')
+plt.suptitle('moving individual source data to fsaverage')
+plt.show()
```
Couldn't this be made more pythonic? Looking and object-oriented MPL should save at least half of the lines for a gain in readability.

TommyClausner replied:
hey @jona-sassenhagen thanks for the feedback! You're totally right ;) - would you like it more like this?

```imgs = [t1_m, t1_s, t1_s]
stcs = [stc, stc, stc_sdr_affine]
titles = ['subject brain', 'fsaverage brain\nprior transformation', 'fsaverage brain\nafter transformation']

fig, ax = plt.subplots(1, 3)
plt.set_cmap('jet')
j = 0
for img, stc, title in zip(imgs, stcs, titles):
    ax[j].imshow(img[slice_sel, :, :] + stc[slice_sel, :, :])
    ax[j].set_title(title)
    ax[j].axis('off')
    j += 1
plt.show()´´´

jona-sassenhagen replied:
Probably best to loop over axes too, while we're at it.

larsoner commented on examples/inverse/plot_morph_stc_to_fsaverage_dipy.py:
```diff
@@ -0,0 +1,231 @@
+==================================================================
+
+This example demonstrates how to morph an individual subject source space to a
+common reference space. For this purpose dipy (http://nipy.org/dipy/) will be
```
do a proper ReSTructuredText link here (look up docs on how or see how we do it elsewhere)

larsoner commented on environment2.yml:
```diff
@@ -26,6 +26,7 @@ dependencies:
 - spyder
 - numexpr
 - pip:
+  - dipy
```
please also add `dipy` to the `doc/conf.py` intersphinx list. Check that it works properly by doing in the `doc` folder` make html_dev-noplot`. You should see it successfully retrieve the `objects.inv` list.

larsoner replied:
If I add to `conf.py` in the `intersphinx_mapping`:
```
    'dipy': ('https://nipy.org/dipy', None),
```
I don't get any warnings with `make html_dev-noplot` other than the standard 5 about images missing (because not all examples were built; plus one about `massich` that is fixed in `master`), and `dipy` links are in the rendered example file. What warnings do you see?

larsoner replied:
Sorry URL is `http` not `https`

TommyClausner replied:
a lot of those:
path/mne-python/doc/auto_examples/time_frequency/plot_compute_csd.rst:17: WARNING: py:func reference target not found: mne.time_frequency.csd_fourier 

and those:
path/mne-python/doc/documentation.rst:: WARNING: toctree contains reference to nonexisting document u'generated/csd_morlet'

TommyClausner replied:
I added exactly what you've added... but I guess that's not the problem anyway...

larsoner replied:
What versions of `sphinx` and `numpydoc` are you using?

TommyClausner replied:
sphinx: 1.7.2
numpydoc: 0.7.0

TommyClausner replied:
oh yes, and I had already http (just claimed it to be "exactly" the same, but this was the only difference ;) )

larsoner replied:
0.7 might be too old, can you try upgrading to see if it helps?

TommyClausner replied:
0.8 seems the newest but yields the same warnings...

larsoner replied:
Okay there must be something wrong with your config we can fix later. In the meantime you can make these changes and CircleCI should built it properly

larsoner commented on examples/inverse/plot_morph_stc_to_fsaverage_dipy.py:
```diff
@@ -0,0 +1,231 @@
+common reference space. For this purpose dipy (http://nipy.org/dipy/) will be
+used in order to perform the necessary transforms to a fsaverage serving as
+reference space. The example uses parts of the MNE example lcmv beamformer
+pipeline (examples/inverse/plot_lcmv_beamformer_volume.py).
```
Should be something like:
```
:ref:`LCMV beamformer pipeline <sphx_glr_auto_examples_inverse_plot_lcmv_beamformer_volume.py>`
```
Try it and see if it renders properly with `make html_dev-noplot`

larsoner commented on examples/inverse/plot_morph_stc_to_fsaverage_dipy.py:
```diff
@@ -0,0 +1,231 @@
+transformation will be applied to the beamformer result. Affine transformations
+are computed based on the mutual information and nonlinear transformations
+will be performed as Symmetric Diffeomorphic Registration using the
+cross-correlation metric [1]_
```
period

TommyClausner replied:
period?

larsoner replied:
at the end of the sentence

TommyClausner replied:
true.

larsoner commented on examples/inverse/plot_morph_stc_to_fsaverage_dipy.py:
```diff
@@ -0,0 +1,231 @@
+cross-correlation metric [1]_
+
+.. warning:: Please do not copy the patterns presented here for your own
+             analysis, this is example is purely illustrative.
```
Is this warning relevant? What here is bad practice?

TommyClausner replied:
nope just was in the example I copied and I just assumed it to be some default reassuring thing since it's an example... I'll remove

larsoner commented on examples/inverse/plot_morph_stc_to_fsaverage_dipy.py:
```diff
@@ -0,0 +1,231 @@
+# uncomment the line(s) below if you encounter problems with ssl certification
+# import ssl
+# if hasattr(ssl, '_create_unverified_context'):
+#     ssl._create_default_https_context = ssl._create_unverified_context
```
Should not need this...?

TommyClausner replied:
well the thing is, in theory it shouldn't be necessary, however I was not able to retrieve the example data without ssl verification ... so in my case it was not possible otherwise... Could also just remove it if you like... (?)

larsoner replied:
yes you shouldn't need it, this is a problem with your Python presumably. Are you on an old 2.7 env?

TommyClausner replied:
actually it's 2.7.15

larsoner commented on examples/inverse/plot_morph_stc_to_fsaverage_dipy.py:
```diff
@@ -0,0 +1,231 @@
+
+# Setup for reading the raw data
+raw = mne.io.read_raw_fif(raw_fname, preload=True)
+raw.info['bads'] = ['MEG 2443', 'EEG 053']  # 2 bads channels
```
bad channels

larsoner commented on examples/inverse/plot_morph_stc_to_fsaverage_dipy.py:
```diff
@@ -0,0 +1,231 @@
+# Read forward model
+forward = mne.read_forward_solution(fname_fwd)
+
+# Compute weights of free orientation (vector) beamformer with weight
```
add a `#################################` (80 char) line before this so sphinx-gallery renders it as a separate section / narrative?

larsoner commented on examples/inverse/plot_morph_stc_to_fsaverage_dipy.py:
```diff
@@ -0,0 +1,231 @@
+# Load nifti data
+
+# load lcmv inverse
+stc = nib.load('lcmv_inverse.nii.gz')
```
let's not call this `stc` -- typically this name is reserved for some variant of `SourceEstimate` and here it is a `NIfTI` object (or whatever `nibabel` calls it).

TommyClausner replied:
hm... any suggestion? nibabel calls their example niftis img... But that's already used by the T1s... what about one of the following:
- src
- img_fct
- img_src
?

agramfort replied:
I would use `img` and call the t1 `t1_img`

TommyClausner replied:
just seen that I'm stupid...
the T1s are called t1_m or t1_s to refer to moving or static, so I could call the source volume img would this be fine? or would you prefer t1_img_m and t1_img_s?

larsoner commented on examples/inverse/plot_morph_stc_to_fsaverage_dipy.py:
```diff
@@ -0,0 +1,231 @@
+t1_s = t1_s.astype('float') / t1_s.max()
+
+##############################################################################
+# Compute affine transformation using mutual information metric
```
can you mention why this metric (it's standard, typically least error, etc.)

larsoner replied:
It's best to have one `References` section -- I'd move the `References` to its own section at the bottom

larsoner commented on examples/inverse/plot_morph_stc_to_fsaverage_dipy.py:
```diff
@@ -0,0 +1,231 @@
+# plot result
+slice_sel = 100
+
+fig = plt.figure()
```
can we use a standard `nilearn` plotter instead? See `nilearn` docs for details.

larsoner commented:
Also please add this example to `doc/documentation.rst` in the "Source-level analysis" section

jona-sassenhagen commented on examples/inverse/plot_morph_stc_to_fsaverage_dipy.py:
```diff
@@ -0,0 +1,242 @@
+cross-correlation metric [1]_.
+
+.. note:: This example does quite a bit of processing, so even on a
+          fast machine it can take a couple of minutes to complete.
```
no reasonable way to speed it up, e.g. by downsampling heavily?

TommyClausner replied:
well, there are two ways to speed it up "reasonably", maybe also a bit depending on how accurate it has to be in the end. E.g. given that the blobs produced by an inverse solution are normally significantly larger than 1 voxel, we could a) downsample as you suggested (maybe by a factor 3 or even 5) and b) reduce the number of iterations, when doing the registration. I just didn't play around too much with the parameters yet (mostly all defaults from all examples), and as it is in the moment it would be a good working high resolution default. But sure: I'll explore the parameters anyway, so we could decide whether we have a fast or accurate solution (ideally both I suppose ;) )

jona-sassenhagen replied:
For the examples, we often lean heavily towards performance on that trade-off. Can just put a note up saying, "for real data, you would" blah blah.

TommyClausner replied:
Your wish is my command! ;)

agramfort commented on doc/documentation.rst:
```diff
@@ -431,6 +431,7 @@ There are also **examples**, which contain a short use-case to highlight MNE-fun
     auto_examples/inverse/plot_mne_crosstalk_function.rst
     auto_examples/inverse/plot_mne_point_spread_function.rst
     auto_examples/inverse/plot_morph_data.rst
+    auto_examples/inverse/plot_morph_stc_to_fsaverage_dipy.rst
```
plot_morph_stc_to_fsaverage_dipy -> plot_morph_stc_to_fsaverage_volume

agramfort commented on examples/inverse/plot_morph_stc_to_fsaverage_dipy.py:
```diff
@@ -0,0 +1,298 @@
+    # compute mapping
+    mapping = sdr.optimize(img_s, img_m_affine)
+
+    return mapping, affine
```
is there a way to save these mapping and affine objects to disk?

agramfort commented on examples/inverse/plot_morph_stc_to_fsaverage_dipy.py:
```diff
@@ -0,0 +1,298 @@
+voxel_size = (3., 3., 3.)  # consider changing for real data scenario
+
+# load lcmv inverse
+img_vol = nib.load('lcmv_inverse.nii.gz')
```
maybe you can avoid saving to disk here and make compute_lcmv_example_data return an img

agramfort commented on examples/inverse/plot_morph_stc_to_fsaverage_dipy.py:
```diff
@@ -0,0 +1,298 @@
+# load lcmv inverse
+img_vol = nib.load('lcmv_inverse.nii.gz')
+
+# reslice lcmv inverse
```
can you make the next code block as function too?

TommyClausner commented:
Hey together,

I can't find out why CircleCI failed... Any idea would be really appreciated!

Thanks in advance ;)

Cheers,
Tommy

larsoner commented:
`Too long with no output (exceeded 10m0s)`

How long does it take to run on your system?

TommyClausner commented:
no longer than ~2min

larsoner commented:
CircleCI is slower than local processing (a factor of 5 or 10x slower isn't unreasonable). Can you try using something much faster (even if the quality is low) and see if it passes?

larsoner commented:
Travis was failing on all PRs for a few days, but appears to be working again. So it was probably some upstream library change/fix that made it work again.

agramfort commented:
why is this not rendered:

https://8256-1301584-gh.circle-artifacts.com/0/html/auto_examples/inverse/plot_morph_stc_to_fsaverage_volume.html

?

TommyClausner commented:
err... is it not? - is there something I could understand wrong about "rendered" ? - because when I click the link that you posted I see a rendered webpage, just as for the other doc pages...

agramfort commented:
yes I meant there is no plots indeed...

TommyClausner commented:
thing is: I know. And the reason why I did it this way (saving, not displaying) is that in the nilearn doc they explicitly emphasized to always close the display... - shall I change it?

larsoner commented:
I would try it locally to see if it's a problem in terms of memory or CPU. If it seems okay, let's see if CircleCI can do it. We want to have informative plots with each example as much as possible.

TommyClausner commented:
hm I see - I'll check ;)
shall I then (case it works) keep the saving or remove it entirely?

agramfort commented:
don't save images to disk but make them visible in the rendered doc

larsoner commented:
Usually with saving we comment out the line and say `# You could save the image to disk with:`

larsoner commented:
Looks reasonable to me, and runs in under 1 min:

https://8278-1301584-gh.circle-artifacts.com/0/html/auto_examples/inverse/plot_morph_stc_to_fsaverage_volume.html

Can you try tweaking the last plots so they look a bit better? Instead of looping you might just want to break it into 3 sphinx-gallery sections (with ######` comments) so each plot gets shown separately instead of being grouped. Or you could try making a single figure with many axes but this might not look as good.

TommyClausner commented:
wasn't the loop to avoid separate plotting in the first place (see way above)? 😛 
but got it: just make nice plots ;)

TommyClausner commented:
hm... it's better now (even though "pretty" is still not the right word I guess)... - do you think it's alright nevertheless?

larsoner commented:
Can you share a link to the CircleCI output?

TommyClausner commented:
err like that? : [https://circleci.com/gh/mne-tools/mne-python/8284?utm_campaign=vcs-integration-link&utm_medium=referral&utm_source=github-build-link](https://circleci.com/gh/mne-tools/mne-python/8284?utm_campaign=vcs-integration-link&utm_medium=referral&utm_source=github-build-link)

larsoner commented:
No in the artifacts, the rendered version of your HTML as I posted above

TommyClausner commented:
how do I get the link? (just modified yours)[https://8284-1301584-gh.circle-artifacts.com/0/html/auto_examples/inverse/plot_morph_stc_to_fsaverage_volume.html](https://8284-1301584-gh.circle-artifacts.com/0/html/auto_examples/inverse/plot_morph_stc_to_fsaverage_volume.html)

larsoner commented:
Log into CircleCI with GitHub. Go to status page for this PR / commit. Artifacts. Navigate to changed file

TommyClausner commented:
thanks! logging in did the trick ;)

larsoner commented:
I think that the timeout is unrelated to your changes

TommyClausner commented:
hm... Other than the Travis timeout it seems fine, right? Is there something we (I) can do about or is it just a matter of time?

Have you got any other things that you would like me to change for the "final version"?

Things I could imagine that are potential candidates for this:

- I didn't use a file extension when saving / loading the mapping (mostly because I didn't know if you'd like .fif and if so if I should keep the filename (``data_path + '/subjects/morph-maps/subject-fsaverage-morph``)), because then it would overwrite the existing file in the example folder... Without file extension it wouldn't though
- I compute the slices to show for the moved volume by multiplying the slice coordinate selected for the Static image with the inverse affine transform. This gives me the more or less "correct" corresponding slices in the different space. However I thought this could also be confusing and I just hard code them... (``slices_m = tuple(np.round(np.matmul(affine.affine_inv, (slices_s + (1,))))[:-1])`` to ``slices_m = (-5, -22, -3)`` given that ``slices_s = (-10, 0, 0)``)?
- I have this if-statement in the mapping-loading-function that checks which type of class is in the list. However this would be unnecessary if the order would stay guaranteed the same (would crash otherwise). Don't know which kind of API you have in mind for later, but if ``mapping`` and ``affine`` are internal variables of e.g. stc in the end and loading and saving stc methods, then we probably wouldn't need this check
- before the plotting I select a "random" volume of the time series to show... Since there is only one volume left due to ``stc.crop(0.0, 0.0)`` which was done to speed up computation time, it's a bit pointless to randomly select of a set of one volume. So I could just generally hard code the first volume... The reason for the random selection was to have a better way of showing what ``index_img()`` does (because a hard coded 0 is not exactly meaningful) - ?
- increase the effort in the plotting?

I mean we don't have to change those things, but given what you are usually commenting on I think those would be worth pointing out ;)

larsoner commented:
> Is there something we (I) can do about or is it just a matter of time?

Someone needs to take the time to figure out why `test_cov` is consistently failing for us now. It will probably be a pain and your time is better spent focusing on the GSoC goals.

> I didn't use a file extension when saving / loading the mapping (mostly because I didn't know if you'd like .fif and if so if I should keep the filename (data_path + '/subjects/morph-maps/subject-fsaverage-morph)), because then it would overwrite the existing file in the example folder.

Does this depend on the source space used, or only on the MRIs? If the latter, then a single vol morph should work, as it's solely MRI structural dependent. If the former, then they shouldn't got in `subjects/` anywhere, since it's analysis-dependent (choice of spacing, etc.).

Is the file in FIF format? If so it should have `.fif` extension.

We should choose a different name in any case, as the `morph map` is specific to Freesurfer surfaces. Probably `volume_morph` or `vol_morph` or so.

> However I thought this could also be confusing and I just hard code them... (slices_m = tuple(np.round(np.matmul(affine.affine_inv, (slices_s + (1,))))[:-1]) to slices_m = (-5, -22, -3) given that slices_s = (-10, 0, 0))?

We will want to create a convenience wrapper for this for people, even though it's a one-liner it's too complex for users. Also `np.dot` is a much more common choice than `matmul`.

How about `morph_slice_indices` that takes an ndarray of shape `(..., 3)` of indices in the source volume and returns an array of the same shape corresponding to slices in the destination volume? You will want to use `einsum` for this rather than `np.dot` or `np.matmul` because you can use `...` in the conversion.

> I have this if-statement in the mapping-loading-function that checks which type of class is in the list. 

I will have to look at the code, but it sounds like passing a list and mangling the types is a recreation of function `arg`/`kwarg` functionality here.

> before the plotting I select a "random" volume of the time series to show... Since there is only one volume left due to stc.crop(0.0, 0.0) which was done to speed up computation time, it's a bit pointless to randomly select of a set of one volume

For `sample` we usually use the auditory condition and use the peak time, see:

https://martinos.org/mne/dev/auto_tutorials/plot_mne_dspm_source_localization.html

> increase the effort in the plotting?

I think they look acceptable now, but if you want to, yes feel free. I think `plot_glass_brain` looks nice.

TommyClausner commented:
All right, thanks for all the suggestions! I implemented (most of) them in the following way:
- data will be saved in the newly created folder subjects/LCMV-results and will contain the morphed volume of the LCMV result and the morph map that has now just .h5 as extension and is called volume_morph
- made a function that applies an affine transform to a N x 3 array using einsum ( <3 )
- I changed the saving / loading a bit to be more elegant, but in general I don't know / think that arg/kwarg is applicable here... depends on your preference maybe... tried some variations and my second self (the user) liked it most, just giving the function [mapping, affine] 
- changed the plot to peak time (0.087ms, right?) and made it nicer
Would you like me to push it would you also have other things that I could implement before doing so?

larsoner commented on examples/inverse/plot_morph_stc_to_fsaverage_volume.py:
```diff
@@ -0,0 +1,404 @@
+    # dissolve object structure
+    for d in data:
+        # save type for order independent decomposition
+        out[type(d).__name__] = d.__dict__
```
Doing things with `__` attributes is generally not a good idea if it can otherwise be avoided.

Here you have two objects, you can explicitly just pass them (same with loading -- you know what should be in the dict returned by `load_mapping` file, just use them directly).

TommyClausner commented:
Well the things is that we didn't really agree on a fix procedure. It somehow evolved to become what it is now :D - so what @larsoner means would be to implement the final API and build the example around it, right? (+tests+doc) and then merge the whole everything, right?
I mean from my perspective I would just do it like this and push every now and then for you to see if it's going in the right direction?
Because then I'd indeed like to build the stc.morph function in a way, that the user always gets the correct morph depending on the underlying data(?) - thoughts?

larsoner commented:
@agramfort what would make it easiest for you?

agramfort commented:
my feeling is that Tommy needs to give it a try. I feel Tommy you are
a bit "paralyzed" by
maybe a lack of clear API guidance. So please try and see how it looks !

show us some ambitious code !

TommyClausner commented:
Hey together :)
I'll make a new PR soon, but I'd like to ask a bunch of things first. 

First what I wanted to do: I want to implement the same functionality as for surface morphs, just for volume morphs. One problem is, that the whole API in ``source_estimate.py`` and ``source_space.py`` is tailored towards surface representations. Which is a problem in the sense, that it becomes somewhat ugly, when just programming around it (even though I found quite elegant solutions so far I hope) or the API of the overlapping functionality would have to be adjusted in general. I mean so far it works, but I think it would be nicer, when common functions like ``morph_data_precomputed`` would accept any input and output correspondingly instead of being actually made for input A and B is just a hack. Long story short: how free am I to change existing functionality (e.g. changing crucial variable names like ``vertices_to``, that the user would have eventually to change compared to previous versions)?

Second just what I did: I added ``stc.morph``, ``stc.morph_precomputed`` to ``VolSourceEstimate``and added or modified the underlying functions (``morph_data_precomputed``, ``compute_morph_sdr``) in ``source_estimate.py``. So I'm technically able to make the output of ``stc.morph`` a *Nifti1Image* or a *VolSourceEstimate* in the morphed space. Now the thing is, then the actual source space (src) wouldn't match stc anymore and ``stc.as_volume`` would for instance not work anymore. And while I'm sure I could make a new interpolator, the surface-related stuff would still be wrong. We could also go for making the output a *Nifti1Image*, but that's kinda lame... I would like 
```
stc_morphed=stc.morph(...)
src_morphed=src.morph(...)
img_morphed=stc_morphed.as_volume(src_morphed)
```
to make the NiftI 

So can I somehow *update* src in order to match the space of the morphed stc?

Another small thing: how do I make ``dipy`` a relative import? Will it automatically magically be found (couldn't test it yet, because I haven't made a test script for the actual modified ``source_estimate.py`` yet)?

I'll make a new PR after knowing what to do about the output and make a somewhat proper-ish test-script ;)

Cheers,
Tommy

agramfort commented:
>
> First what I wanted to do: I want to implement the same functionality as
> for surface morphs, just for volume morphs. One problem is, that the whole
> API in source_estimate.py and source_space.py is tailored towards surface
> representations.
>
who said it would be trivial :)

> Which is a problem in the sense, that it becomes somewhat ugly, when just
> programming around it (even though I found quite elegant solutions so far I
> hope) or the API of the overlapping functionality would have to be adjusted
> in general. I mean so far it works, but I think it would be nicer, when
> common functions like morph_data_precomputed would accept any input and
> output correspondingly instead of being actually made for input A and B is
> just a hack. Long story short: how free am I to change existing
> functionality (e.g. changing crucial variable names like vertices_to,
> that the user would have eventually to change compared to previous
> versions)?
>
we should try to avoid changing API of public functions. I don't think
vertices_to
is exposed to users. I approach the pb with private functions. If the code
is like this now:

def foobar(input):
   blabla

to:

def _foobar_surf(input):
    blabla

def _foobar_vol(input):
    blabla

def foobar(input):
   if input is surf:
       return _foobar_surf(input)
   else:
       return _foobar_vol(input)

makes sense?

> Second just what I did: I added stc.morph, stc.morph_precomputed to
> VolSourceEstimateand added or modified the underlying functions (
> morph_data_precomputed, compute_morph_sdr) in source_estimate.py. So I'm
> technically able to make the output of stc.morph a *Nifti1Image* or a
> *VolSourceEstimate* in the morphed space. Now the thing is, then the
> actual source space (src) wouldn't match stc anymore and stc.as_volume
> would for instance not work anymore. And while I'm sure I could make a new
> interpolator, the surface-related stuff would still be wrong. We could also
> go for making the output a *Nifti1Image*, but that's kinda lame... I
> would like
>
> stc_morphed=stc.morph(...)
> src_morphed=src.morph(...)
> img_morphed=stc_morphed.as_volume(src_morphed)
>
> to make the NiftI
>
yes. Sorry I am a bit lost about the details without seeing the code and
you pointing me to a line
that creates the issue.

> So can I somehow *update* src in order to match the space of the morphed
> stc?
>
> Another small thing: how do I make dipy a relative import?
>
just import dipy from inside the function.

> Will it automatically magically be found (couldn't test it yet, because I
> haven't made a test script for the actual modified source_estimate.py
> yet)?
>
> I'll make a new PR after knowing what to do about the output and make a
> somewhat proper-ish test-script ;)
>
you can make a WIP PR asap so we can see what the low level issues are.
